src/session_impl.cpp

```diff
@@ -6545,24 +6591,44 @@ namespace {
 	void session_impl::set_external_address(address const& ip
 		, int const source_type, address const& source)
 	{
-#ifndef TORRENT_DISABLE_LOGGING
-		if (should_log())
-		{
-			session_log(": set_external_address(%s, %d, %s)", print_address(ip).c_str()
-				, source_type, print_address(source).c_str());
-		}
-#endif
-
 		// for now, just pick the first socket with a matching address family
-		// TODO: 3 allow the caller to select which listen socket to update
+		// TODO: remove this function once all callers are updated to specify a listen socket
 		for (auto& i : m_listen_sockets)
 		{
 			if (i.local_endpoint.address().is_v4() != ip.is_v4())
 				continue;
 
-			if (!i.external_address.cast_vote(ip, source_type, source)) return;
+			set_external_address(i, ip, source_type, source);
 			break;
 		}
+	}
+
+	void session_impl::set_external_address(
+		tcp::endpoint const& local_endpoint, address const& ip
+		, int const source_type, address const& source)
+	{
+		for (auto& sock : m_listen_sockets)
```

Comment:
unless this is placeholder logic, do you think it would be simpler to use ``std::find_if()`` followed by the call to ``set_external_address()``?